src/server/Server.java

```diff
@@ -1,214 +1,199 @@
+    }
+
+    public void setWWWDirectory(String in) {
+        WWWDir = in;
+    }
+
     public void setPort(int in) {
         Port = in;
     }
 
-	public void addCustomHeader(Entry<String, String> in){
-		CustomHeaders.put(in.getKey(), in.getValue());
-	}
-
-	public void setMaximumSocketTries(int in) {
-		MaxTries = in + 1;
-	}
+    public void addCustomHeader(Entry<String, String> in) {
+        CustomHeaders.put(in.getKey(), in.getValue());
+    }
 
-    public void HTTPStart() throws Exception {
-		loadMIME();
-        ExecutorService executor = Executors.newFixedThreadPool(MaxConcurrentRequests);
-        try (ServerSocket SS = new ServerSocket(Port, backlog)) {
-			log.s("Server running at :"+Port);
-            while (true) {
-                /*
-                 * Max x retries (1ms delay between each) for every request to process if
-                 * MaxConcurrentRequests is reached
-                 */
-                int tries = 0; // current tries
-                inner: while (tries < MaxTries) {
-                    if (tries > 0) Thread.sleep(1);
-                    if (CurrentConcurrentRequests <= MaxConcurrentRequests) {
-                        Socket S = SS.accept();
-                        S.setKeepAlive(false);
-						S.setTcpNoDelay(true);
-						S.setReceiveBufferSize(BufferSize);
-						S.setSendBufferSize(BufferSize);
-						S.setSoTimeout(SOTimeout);
-                        executor.execute(new Engine(S));
-                        CurrentConcurrentRequests++;
-                        break inner;
-                    } else tries++;
+    private void mainLoop(ServerSocket serverSocket, ThreadPoolExecutor poolExecutor) {
+        while (true) {
+            try {
+                Socket s = serverSocket.accept();
+                s.setKeepAlive(false);
+                s.setTcpNoDelay(true);
+                s.setReceiveBufferSize(BufferSize);
+                s.setSendBufferSize(BufferSize);
+                s.setSoTimeout(SOTimeout);
+
+                try {
+                    poolExecutor.execute(new Engine(s));
+                } catch (RejectedExecutionException ignore) {
+                    log.i("concurrent connections exceed the configured maximum");
+                    Network.write(new BufferedOutputStream(s.getOutputStream()), new byte[]{}, new byte[]{}, HTTPCode.SERVICE_UNAVAILABLE.getBytes(), false, new HashMap<>(), false);
                 }
+
+            } catch (IOException e) {
+                log.e(e.getMessage());
+                return;
             }
         }
-        finally {
-            executor.shutdownNow();
+    }
+
+    public void HTTPStart() throws Exception {
+        loadMIME();
+        final int nCores = Runtime.getRuntime().availableProcessors();
+        ArrayBlockingQueue<Runnable> workQueue = new ArrayBlockingQueue<>(MaxConcurrentRequests);
+        ThreadPoolExecutor poolExecutor = new ThreadPoolExecutor(nCores, nCores * 2, 60, TimeUnit.SECONDS, workQueue);
```

Comment:
Re-initializing threads after being idle for 60 seconds can add some latency for new requests, check [This](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ThreadPoolExecutor.html#%3Cinit%3E(int,int,long,java.util.concurrent.TimeUnit,java.util.concurrent.BlockingQueue)). Is it a tradeoff تستاهل?

Reply:
It's only re-initializing the threads that are dynamically created to reach `maximumPoolSize` (`nCores * 2`), `nCores` threads are always running.

This allows a little more requests (`nCores` requests) to be accepted in case the queue is full, the other option is to make it a fixed pool.

Reply:
Make it a fixed pool, that would increase performance